src/main/java/org/broadinstitute/hellbender/tools/spark/validation/CompareDuplicates.java

```diff
@@ -0,0 +1,192 @@
+        if (firstDupesCount != secondDupesCount) {
+            System.out.println("BAMs have different number of total duplicates: " + firstDupesCount + "," + secondDupesCount);
+        }
+        System.out.println("first and second: " + firstDupesCount + "," + secondDupesCount);
+
+        Broadcast<SAMFileHeader> bHeader = ctx.broadcast(getHeaderForReads());
+        // Group the reads of each BAM by MarkDuplicates key, then pair up the the reads for each BAM.
+        JavaPairRDD<String, GATKRead> firstKeyed = firstReads.mapToPair(read -> new Tuple2<>(ReadsKey.keyForFragment(bHeader.getValue(), read), read));
+        JavaPairRDD<String, GATKRead> secondKeyed = secondReads.mapToPair(read -> new Tuple2<>(ReadsKey.keyForFragment(bHeader.getValue(), read), read));
+        JavaPairRDD<String, Tuple2<Iterable<GATKRead>, Iterable<GATKRead>>> cogroup = firstKeyed.cogroup(secondKeyed);
+
+
+        // Produces an RDD of MatchTypes, e.g., EQUAL, DIFFERENT_REPRESENTATIVE_READ, etc. per MarkDuplicates key,
+        // which is approximately start position x strand.
+        JavaRDD<Utils.MatchType> tagged = cogroup.map(v1 -> {
+            SAMFileHeader header = bHeader.getValue();
+
+            Iterable<GATKRead> iFirstReads = v1._2()._1();
+            Iterable<GATKRead> iSecondReads = v1._2()._2();
+
+            return Utils.getDupes(iFirstReads, iSecondReads, header);
+        });
+
+        // TODO: We should also produce examples of reads that don't match to make debugging easier (#1263).
+        Map<Utils.MatchType, Integer> tagCountMap = tagged.mapToPair(v1 ->
+                new Tuple2<>(v1, 1)).reduceByKey((v1, v2) -> v1 + v2).collectAsMap();
+
+        if (tagCountMap.get(Utils.MatchType.SIZE_UNEQUAL) != null) {
+            throw new UserException("The number of reads by the MarkDuplicates key were unequal, indicating that the BAMs are not the same");
+        }
+        if (tagCountMap.get(Utils.MatchType.READ_MISMATCH) != null) {
+            throw new UserException("The reads grouped by the MarkDuplicates key were not the same, indicating that the BAMs are not the same");
+        }
+        if (printSummary) {
+            Utils.MatchType[] values = Utils.MatchType.values();
+            HashSet<Utils.MatchType> matchTypes = Sets.newHashSet(values);
+            System.out.println("##############################");
+            matchTypes.forEach(s -> System.out.println(s + ": " + tagCountMap.getOrDefault(s, 0)));
+        }
+
+        if (throwOnDiff) {
+            for (Utils.MatchType s : Utils.MatchType.values()) {
+                if (s != Utils.MatchType.EQUAL) {
+                    if (tagCountMap.get(s) != null)
+                    throw new UserException("found difference between the two BAMs: " + s + " with count " + tagCountMap.get(s));
+                }
+            }
+        }
+    }
+
+    private static class Utils implements Serializable {
```

Comment:
I think this class should be inlined since it's just static functions and an enum.